### PR TITLE
Moving to hpke-rs

### PR DESCRIPTION
**Pull request by armfazh**
Moving away from `hpke` to `hpke-rs` crate.

### CODE REVIEW

**Review by bhalleycf — APPROVED** [review]

File: daphne/src/hpke.rs [hunk 1]
```diff
@@ -113,66 +110,18 @@ pub struct HpkeReceiverConfig {
...
-                // We should never get here.
-                unreachable!("Unsupported KEM ({:?})", self.config.kem_id);
-            }
-        }
+        let sender: Hpke<ImplHpkeCrypto> =
+            check_suite(self.config.kem_id, self.config.kdf_id, self.config.aead_id)?;
+        let pk = HpkePublicKey::new(self.config.public_key.clone());
+        let (enc, mut ctx) = sender.setup_sender(&pk, info, None, None, None)?;
+        let ciphertext = ctx.seal(aad, plaintext)?;
+        Ok((enc, ciphertext))
```
Comment (cjpatton) [inline daphne/src/hpke.rs:124]: :D
```suggestion
        self.config.encrypt(info, aad, plaintext)
```

File: daphne/src/hpke.rs [hunk 2]
```diff
@@ -184,75 +133,39 @@ impl HpkeReceiverConfig {
...
-                }
-            }
-            _ => {
-                // We should never get here.
-                unreachable!("Unsupported KEM ({:?})", kem_id);
-            }
+        let kem = match kem_id {
+            HpkeKemId::P256HkdfSha256 => KemAlgorithm::DhKemP256,
+            HpkeKemId::X25519HkdfSha256 => KemAlgorithm::DhKem25519,
+            _ => unreachable!("Unsupported KEM ({:?})", kem_id),
```
Comment (cjpatton) [inline daphne/src/hpke.rs:149]: Matching "_" for enumerated types is a bit of a footgun in my experience. Better to be explicit!

```suggestion
            NotImplemented(x) => unreachable!("Unsupported KEM ({:?})", x),
```

File: daphne/src/hpke.rs [hunk 3]
```diff
@@ -184,75 +133,39 @@ impl HpkeReceiverConfig {
...
+                    id,
+                    kem_id,
+                    kdf_id: HpkeKdfId::HkdfSha256,
+                    aead_id: HpkeAeadId::Aes128Gcm,
+                    public_key: Vec::from(pk.as_slice()),
+                },
+                secret_key: Vec::from(sk.as_slice()),
+            };
+        } else {
+            unreachable!("bad key generation for KEM ({:?})", kem_id)
```
Comment (cjpatton) [inline daphne/src/hpke.rs:168]: Is this actually "unreachable"? Why not propagate the error here?

**Review by cjpatton — CHANGES_REQUESTED** [review]
Nice! Just a few comments.

File: daphne/src/hpke.rs (hunk 1 above)
Comment (armfazh) [reply daphne/src/hpke.rs]: so true,

File: daphne/src/hpke.rs (hunk 2 above)
Comment (armfazh) [reply daphne/src/hpke.rs]: done

File: daphne/src/hpke.rs (hunk 3 above)
Comment (armfazh) [reply daphne/src/hpke.rs]: take a look

File: daphne/src/lib.rs [hunk 4]
```diff
@@ -378,7 +372,10 @@ impl DapGlobalConfig {
         let kem_ids = self.supported_hpke_kems.clone();
         kem_ids.into_iter().enumerate().map(move |(i, kem_id)| {
             let (config_id, _overflowed) = first_config_id.overflowing_add(i as u8);
-            HpkeReceiverConfig::gen(config_id, kem_id)
+            match HpkeReceiverConfig::gen(config_id, kem_id) {
+                Ok(cfg) => cfg,
+                Err(e) => panic!("{:?}", e),
```
Comment (armfazh) [inline daphne/src/lib.rs:377]: @cjpatton any recommendation here to avoid panic?

File: daphne/src/hpke.rs [hunk 5]
```diff
@@ -38,44 +68,26 @@ impl HpkeConfig {
...
+    }
 
-                Ok((enc.to_bytes().to_vec(), sender.seal(plaintext, aad)?))
-            }
-            _ => {
-                // We should never get here.
-                panic!("Unsupported KEM ({:?})", self.kem_id);
-            }
-        }
+    pub fn decrypt(
```
Comment (cjpatton) [inline daphne/src/hpke.rs:78]:
```suggestion
    pub(crate) fn decrypt(
```

**Review by cjpatton — APPROVED** [review]
Looks good, one minor change.

File: daphne/src/lib.rs [hunk 6]
```diff
@@ -378,7 +372,10 @@ impl DapGlobalConfig {
         let kem_ids = self.supported_hpke_kems.clone();
         kem_ids.into_iter().enumerate().map(move |(i, kem_id)| {
             let (config_id, _overflowed) = first_config_id.overflowing_add(i as u8);
-            HpkeReceiverConfig::gen(config_id, kem_id)
+            match HpkeReceiverConfig::gen(config_id, kem_id) {
```
Comment (cjpatton) [inline daphne/src/lib.rs:375]: No need to match, you could just unwrap() or expect():
```rust
     HpkeReceiverconfig::gen(config_id, kem_id).expect("failed to generate HPKE receiver config")
```

File: daphne/src/lib.rs (hunk 4 above)
Comment (cjpatton) [reply daphne/src/lib.rs:377]: I would probably just propagate the DapError here. I think it's fine to panic, as long as the only way to panic is an RNG failure?

File: daphne/src/hpke.rs (hunk 5 above)
Comment (armfazh) [reply daphne/src/hpke.rs]: done

File: daphne/src/lib.rs (hunk 4 above)
Comment (armfazh) [reply daphne/src/lib.rs]: it is possible that kem_id is invalid, so better to propagate error.

**Review by cjpatton — APPROVED** [review]

**Pull request merged**